Switch to USM Host Memory for Reduction Result

julianmi opened the pull request:
This patch aims at reducing the required transfer times in the family of reduce algorithms for copying the result back to the host.

The current mainline version uses a single value buffer that is wrapped in a future-like object. This patch introduces a future-like class for reduction results based on USM host memory. This eliminates the L0 call to `zeCommandListAppendMemoryCopy` and directly transfers the result instead. This improves the transfer times significantly with great runtime benefits for smaller input arrays.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +573,45 @@ class __future : private std::tuple<_Args...>
+    {
+        using _Tp = typename ::std::remove_pointer<_Res>::type;
+        __my_event.wait();
+        _Tp __local_val = __my_res[0];
```
Why we can't use ```auto``` here instead of ```_Tp``` type ?

julianmi replied:
I think we can. I will change it.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +573,45 @@ class __future : private std::tuple<_Args...>
+        using _Tp = typename ::std::remove_pointer<_Res>::type;
+        __my_event.wait();
+        _Tp __local_val = __my_res[0];
+        sycl::free(__my_res, __my_exec.queue());
```
If you call ```free``` function not in destructor, I think possible cases when the instance of this class will be destroyed without ```free``` call.

julianmi replied:
Good point, thanks. I will change that.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -604,10 +606,8 @@ class __reduce_future
         __my_event.wait();
-        _Tp __local_val = __my_res[0];
-        sycl::free(__my_res, __my_exec.queue());
+        auto __local_val = __my_res[0];
```
or may be more simple: ```return __my_res[0];```

julianmi replied:
This is what I have just realized and changed.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +573,44 @@ class __future : private std::tuple<_Args...>
+    }
+
+    auto
+    get()
```
May we declare this function as ```const``` ?

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +573,44 @@ class __future : private std::tuple<_Args...>
+    _Res __my_res;
+
+  public:
+    __reduce_future(_ExecutionPolicy __exec, _Event __e, _Res __res)
```
Usually you construct the instance of this class like
```__reduce_future(::std::forward<_ExecutionPolicy>(__exec), ...```
but this declaration mean that move semantic here really isn't possible, I think.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +574,48 @@ class __future : private std::tuple<_Args...>
+
+  public:
+    __reduce_future(const _ExecutionPolicy& __exec, _Event __e, _Res __res)
+        : __my_exec(__exec), __my_event(__e), __my_res(::std::make_shared<_Res>(__res))
```
Since `_Event __e` is passed by value in the constructor, in the initializer list it may be better to have `__my_event(::std::move(__e))` to avoid an unnecessary copy.

julianmi replied:
Agreed. I will change that.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +574,48 @@ class __future : private std::tuple<_Args...>
+{
+    _ExecutionPolicy __my_exec;
+    _Event __my_event;
+    ::std::shared_ptr<_Res> __my_res;
```
Should this class be copy constructible or should we allow shared ownership? If we do not need it to be copy constructible, then I think the best option is to delete the copy constructor / assignment operator and make this `::std::unique_ptr<_Res>`

julianmi replied:
I tend to agree that our future-like objects should not be copied similar to `std::future`. I will change that if there are no other opinions on when copying a future could be useful.

SergeyKopienko replied:
Fixed: now we using ::std::unique_ptr<_Res> with custom deleter.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -606,8 +611,8 @@ class __reduce_future
-        __my_event.wait();
-        return __my_res[0];
+        __my_event.wait_and_throw();
+        return *__my_res.get()[0];
```
return (*__my_res)[0];

SergeyKopienko commented:
@julianmi please review and merge the PR https://github.com/oneapi-src/oneDPL/pull/875 into this branch

julianmi commented:
> @julianmi please review and merge the PR #875 into this branch

Those are great improvements, thanks @SergeyKopienko.

SergeyKopienko reviewed (approved):
LGTM

akukanov reviewed (dismissed):
I would like @MikeDvorskiy to also take a look before this patch is committed.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -140,12 +139,13 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
                     {
                         __reduce_pattern.apply_init(__init, __result);
-                        __res_acc[0] = __result;
+                        __res[0] = __result;
```
*__res = __result;

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -269,7 +268,7 @@ struct __parallel_transform_reduce_impl
                             {
                                 __reduce_op.apply_init(__init, __result);
-                                __res_acc[0] = __result;
+                                __res[0] = __result;
```
*__res = __result;

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +575,45 @@ class __future : private std::tuple<_Args...>
 
+// A contract for a future class for reduce: <execution policy, sycl::event, USM host memory for the reduced value>
+template <typename _ExecutionPolicy, typename _Event, typename _Res>
+class __reduce_future
```
I think we don't need the same addition feature type.
We may generalize existing one  - class __future, by  passing instead of buffer some RAII type based on ::std::unique_ptr. f.e.
```
template<typename Res>
using usm_value_buf = ::std::unique_ptr<_Res, ::std::function<void(_Res*)>>;
auto make_usm_value_buf(queue q, Res* res) { return usm_value_buf(res, [q](_Res* __res) { ::sycl::free(res, queue); }); }
```

rarutyun replied:
I don't suggest to have type-erased utility (like `std::function`). Let's avoid allocation we don't control. You can always write deleter as a named class (not as a lambda) so I don't see why this overhead is necessary.

In general I am in favor of genericity idea. We always use either buffer of pointer, right? Can we just have compile-time check and rely that pointer always was allocated via USM?

SergeyKopienko replied:
Usage of ```std::function``` were removed from this PR.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -120,11 +120,10 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
         const _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item);
 
-        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+        _Tp* __res = sycl::malloc_host<_Tp>(1, __exec.queue());
```
Regarding the host memory - what of we have async call and further async calls on the device and we don't need access to reduction result on the host? We should bot get performance degradation in that case.

julianmi replied:
How would you access the result on the device then? Calling `get()` on the returned future would return a host accessor to the data.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -65,19 +65,19 @@ struct __parallel_transform_reduce_seq_submitter<_Tp, __internal::__optional_ker
                 _Tp __result = __transform_pattern(__n, __rngs...);
                 __reduce_pattern.apply_init(__init, __result);
-                __res_acc[0] = __result;
+                __res[0] = __result;
```
Done.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -120,11 +120,10 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
         const _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item);
 
-        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+        _Tp* __res = sycl::malloc_host<_Tp>(1, __exec.queue());
```
Good point. I think this would require a separate buffer and USM kernel which would add quite a bit of code. Is that worth it for the discussed performance improvements?

SergeyKopienko reviewed (approved):
LGTM

akukanov commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +652,20 @@ class __future : private std::tuple<_Args...>
+        return false;
+    if (!__device.has(sycl::aspect::usm_host_allocations))
+        return false;
+    if (__device.get_info<sycl::info::device::vendor_id>() != 32902)
+        return false;
```
Filtering out some code paths based on vendor id instead of HW capabilities might cause bad perception. What alternative checks might be used here?

julianmi replied:
I think the key characteristic is the backend. L0 provides a low-overhead implementation of USM host memory whereas other backends use slower pinned host memory (mlock on Linux).

SergeyKopienko commented:
There are 3 places in code like
```__storage __res_container = __storage<_ExecutionPolicy, _Tp>(__exec, __use_usm, 1);```

but in all these places you pass __res_container  into your Kernel code by reference: ```[&]``` in lambda.
For example, will take a look into
```struct __parallel_transform_reduce_small_submitter```
passing instance of ```__res_container``` by reference, from my point of view this mean that that there are no shared_ptr pointers incrementation doing.

And if I am right, then you return result
```return __future(__reduce_event, __res_container);```
and on caller side we will forget about in, this mean that you container(s) will be delated but still used inside the Kernel code.

It's my opinion.
May be I am wrong but better to check.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+    using __accessor_t = sycl::accessor<_T, 1, sycl::access::mode::read_write, __dpl_sycl::__target_device,
+                                        sycl::access::placeholder::false_t>;
+    __accessor_t m_acc;
+    _T* m_ptr;
```
Better to initialize this field on declaration: ```_T* m_ptr = null_ptr;```

julianmi replied:
Done.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+    using __sycl_buffer_t = sycl::buffer<_T, 1>;
+    ::std::shared_ptr<__sycl_buffer_t> m_sycl_buf;
+    ::std::shared_ptr<_T> m_usm_buf;
+    bool m_usm;
```
Better to initialize this field on declaration: ```bool m_usm = false;```

julianmi replied:
Done.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+        __accessor<_T> acc;
+        if (m_usm)
+        {
+            acc.m_usm = true;
```
We can move this duplicated line outside of condition: ```acc.m_usm = m_usm;```

julianmi replied:
Done.

SergeyKopienko commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+    }
+
+    bool
+    get_usm()
```
I believe we can declare this method as ```const```

julianmi replied:
Done.

SergeyKopienko replied:
I mean
```bool
get_usm() const
{
    return m_usm;
}
```

julianmi replied:
Of course, my bad. I fixed it.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+{
+  private:
+    using __sycl_buffer_t = sycl::buffer<_T, 1>;
+    ::std::shared_ptr<__sycl_buffer_t> m_sycl_buf;
```
Since we are exclusively using `m_sycl_buf` or `m_usm_buf`, it may be worth it to move these two variables into a union or `std::variant` so we only ever store one pointer in this class.

julianmi replied:
This was tried before but rendered tricky when passing them to a kernel. Thus, the current solution uses smart pointers which should incur only minor overheads. The runtime performance is very close to the explicit mode proposed in #1106.

mmichel11 replied:
Ah, I see. This makes sense, then.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+                                        sycl::access::placeholder::false_t>;
+    __accessor_t m_acc;
+    _T* m_ptr;
+    bool m_usm;
```
Done.

julianmi commented:
I've rebased this PR and introduced checks to support multiple compiler versions. Currently, DPC++ 2023.1 and 20230822 and above support the USM host approach proposed here. Other compilers fall back to the existing buffer implementation. I've verified this approach with oneAPI releases back to 2021.4. The patch shows similar performance to the mainline when using buffers and significant speedups when using USM host memory on L0 GPUs, especially for small arrays.

julianmi commented:
> There are 3 places in code like
> 
> ```c++
> __storage __res_container = __storage<_ExecutionPolicy, _Tp>(__exec, __use_usm, 1);
> ```
> 
> but in all these places you pass __res_container into your Kernel code by reference: `[&]` in lambda. For example, will take a look into
> 
> ```c++
> struct __parallel_transform_reduce_small_submitter
> ```
> 
> passing instance of `__res_container` by reference, from my point of view this mean that that there are no shared_ptr pointers incrementation doing.
> 
> And if I am right, then you return result
> 
> ```c++
> return __future(__reduce_event, __res_container);
> ```
> 
> and on caller side we will forget about in, this mean that you container(s) will be delated but still used inside the Kernel code.
> 
> It's my opinion. May be I am wrong but better to check.

Thanks for raising this, @SergeyKopienko. I think this issue could occur if someone calls `reduce_async` without getting the result. I don't think this is a realistic case though. And even if someone would do this, the runtime manages the lifetime of buffers and the USM version is only used with L0, which pools memory operations. I couldn't reproduce this issue with a long-running kernel and USM or buffer memory.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -523,6 +602,16 @@ class __future : private std::tuple<_Args...>
+        // Explicit wait in case of USM memory. Buffer accessors are synchronous.
+        if (__buf.get_usm())
+            wait();
+        return __buf.get_value();
```
SYCL buffer guarantees just data synchronization... But a kernel might still work...

julianmi replied:
I think the data dependence should be on the buffer that is used in the kernel (read/write access). Thus, the host accessor called here should wait for the kernel completion. This follows the existing `__wait_and_get_value(sycl::buffer<_T>& __buf)` implementation above.

MikeDvorskiy replied:
"Thus, the host accessor called here should wait for the kernel completion"

I don't think so. The host accessor is waiting for the data will be copied back to the host, nothing more.

MikeDvorskiy replied:
Actually, looking at the spec:
```
As said in [Section 3.9.8](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:synchronization), the only blocking operations in SYCL (apart from explicit wait operations) are:

host accessor constructor, which waits for any kernels enqueued before its creation that write to the corresponding object to finish and be copied back to host memory before it starts processing. The host accessor does not necessarily copy back to the same host memory as initially given by the user;
```

So,

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -573,6 +662,26 @@ class __future : private std::tuple<_Args...>
+inline bool
+__use_USM_host_allocations(sycl::queue __queue)
+{
+// The unified future above is currently only supported by DPCPP 2023.1 and compilers starting from 20230820.
```
It looks very weird...
Why would not just call `__device.has(sycl::aspect::usm_host_allocations`? or why such call doesn't  make the all necessary checks?
A question rather is to a runtime team...

julianmi replied:
We can query the support for USM host memory with `sycl::aspect::usm_host_allocations`. However, its use has shown to be beneficial on L0 GPUs only. This routine checks for all three conditions and uses guards to fall back to USM for compilers not supporting the unified future implementation.

MikeDvorskiy reviewed (dismissed):
LGTM

mmichel11 reviewed (dismissed):
LGTM. It's probably best to check with @SergeyKopienko as well if he's ready to approve.

SergeyKopienko reviewed (dismissed):
LGTM

akukanov commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,86 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+#else
+    __accessor(sycl::handler& __cgh, bool __usm, ::std::shared_ptr<sycl::buffer<_T, 1>> m_sycl_buf,
+               ::std::shared_ptr<_T> m_usm_buf)
+        : m_usm(__usm), m_acc(sycl::accessor(*m_sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{}))
```
Should not `m_usm` be set to `false` there, as de facto the buffer is used?

julianmi replied:
Done.

akukanov commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,86 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+    bool m_usm = false;
+
+  public:
+// Compilers before 2023.1 don't provide a default constructor for sycl::accessor. Fallback to a buffer instead.
```
The comment talks about a fallback, yet precedes the "main" branch. It should better be moved or rephrased.

julianmi replied:
I've reworded the comment.

akukanov commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:
```diff
@@ -508,8 +509,86 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
+
+  public:
+// Compilers before 2023.1 don't provide a default constructor for sycl::accessor. Fallback to a buffer instead.
+#if __SYCL_COMPILER_VERSION == 20230320 || __SYCL_COMPILER_VERSION >= 20230820
```
Does it make sense to introduce a "feature presence" macro in the config header?

SergeyKopienko reviewed (approved):
LGTM

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -78,6 +78,14 @@
+// The unified future supporting USM host memory and buffers is only supported by DPCPP 2023.1
+// and compilers starting from 20230820.
+#if (__SYCL_COMPILER_VERSION == 20230320 || __SYCL_COMPILER_VERSION >= 20230820)
+#    define _ONEDPL_SYCL_USM_HOST_PRESENT 1
```
I've switched to using `_ONEDPL_LIBSYCL_VERSION` to exclude the problematic compiler 2023.1 which has `_ONEDPL_LIBSYCL_VERSION == 60200`.

SergeyKopienko reviewed (approved):
LGTM
